Feature Associations

RossielCS opened the pull request:
### This PR addresses the associations between User and Event models, through Invitation.
Features:
* Users can register on the website.
* Creation and destruction of sessions.
* Users create events and assist to event.
* Users can invite others to an event.

RossielCS committed:
Generate user model

- Add columns to users
- Validations in user model
- Rubocop and stickler

RossielCS committed:
Add users controller and routes

- New, create and show methods and routes for users
- Add bcrypt gem

RossielCS committed:
Add new and show users views

- App layout loop for flash messages

RossielCS committed:
Add sessions controller

- New, create, destroy sessions methods
- log_in, log_out, current_user methods

RossielCS committed:
Add event model and controller

RossielCS committed:
Generate invitation model and add associations

- Add associations
- create_invitation method
- user can select event attendees

RossielCS committed:
Add upcoming and previous events methods

- Methods in user model
- Scopes in event model
- Partial to show events table

RossielCS committed:
Add links in layouts app

- Links to jump between views
- Fix rubocop offenses

RossielCS committed:
Update README

RossielCS committed:
Update README

RossielCS committed:
Remove cops in rubocop config file

suretrust commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,34 @@
+
+    if @event.save
+      create_invitation(params[:user_ids], @event.id)
+      flash.now[:notice] = 'Event was successfully created.'
```
This should be `flash[:notice]`. Check [here](http://spaghettirefactory.blogspot.com/2015/03/rails-flash-vs-flashnow.html#:~:text=Boiled%20down%20to%20the%20very,page%20after%20clicking%20a%20link.) for more clarification.

suretrust commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,34 @@
+      flash.now[:notice] = 'Event was successfully created.'
+      redirect_to @event
+    else
+      flash.now[:notice] = 'Error.'
```
This should be `flash[:notice]`. Check [here](http://spaghettirefactory.blogspot.com/2015/03/rails-flash-vs-flashnow.html#:~:text=Boiled%20down%20to%20the%20very,page%20after%20clicking%20a%20link.) for more clarification.

suretrust commented on app/helpers/events_helper.rb:
```diff
@@ -0,0 +1,7 @@
+module EventsHelper
+  def create_invitation(attendees, event_id)
+    attendees.each do |attendee|
+      Invitation.create(created_event_id: event_id, attendee_id: attendee)
+    end
+  end
```
This is where the app is breaking from when I try to create an event.

suretrust commented on app/views/events/index.html.erb:
```diff
@@ -0,0 +1,6 @@
+
+<br>
+
+<%= render "users/showtb" %>
```
This partial should be put in events, after all, it contains past and upcoming events. `events/showtb`. You could also improve the naming, `showtb`.

suretrust commented on app/models/event.rb:
```diff
@@ -0,0 +1,12 @@
+class Event < ApplicationRecord
+  validates :description, :location, :date, presence: true
+  validates :description, length: { in: 10..120 }
+  validates :location, length: { in: 5..50 }
```
Since you already validated the length of `description` and `location`, that validation automatically covers for `presence: true`, so you can remove `:description` and `:location` from `L2` to avoid repetition.

suretrust commented on app/models/event.rb:
```diff
@@ -0,0 +1,12 @@
+
+  belongs_to :creator, class_name: 'User'
+  has_many :invitations, foreign_key: 'created_event_id'
+  has_many :attendees, through: :invitations, source: :attendee
```
I'm not sure why you need `source: :attendee`

suretrust commented on app/models/event.rb:
```diff
@@ -0,0 +1,12 @@
+  validates :location, length: { in: 5..50 }
+
+  belongs_to :creator, class_name: 'User'
+  has_many :invitations, foreign_key: 'created_event_id'
```
I'm not sure why you needed to change the foreign key, `foreign_key: 'created_event_id'`. `invitation_id` sounds good already.

suretrust commented on app/models/user.rb:
```diff
@@ -0,0 +1,35 @@
+  has_secure_password
+
+  has_many :created_events, foreign_key: 'creator_id', class_name: 'Event'
+  has_many :invitations, foreign_key: 'attendee_id'
+  has_many :attended_events, through: :invitations, source: :created_event
```
Ditto, what I mentioned in the event model.

suretrust reviewed (changes requested):
Hi @RossielCS, 👋 

### Great job thus far! 🙌 Please see my comments. 👇 

* [x] It has functional `login` page (authentication **WITHOUT** password, just providing a user email). It's cool that you can achieve full auth with password, but can you also build this app without it? ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/login_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] You app breaks when I try to create a new event.
<img width="1075" alt="Screenshot 2020-06-13 at 07 06 14" src="https://user-images.githubusercontent.com/31304943/84561464-8640da80-ad44-11ea-8610-0f4031a76771.png">

- [x] Navigation links should be present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

**Kindly submit for another CRR once these issues have been fixed.** _Happy coding!_ 🎉

RossielCS committed:
Apply request changes

- Change associations and flash msgs
- Navigations links at top
- Remove password

Anna-Myzukina commented:
## Hi @RossielCS ! Very nice project! :wave: 

# You did a great job so far.

### A few changes before it can be approved :eyes:  
#### A database schema should be similar to:


#### Associations

##### User as event creator

* [ ] User model: `has_many :events, foreign_key: :creator_id`  from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)
![screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) 
* [ ] Event model `belongs_to :creator, class_name: "User"`  from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)
![screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png)
  * Migrations
    * [ ] events table has creator_id as the foreign key  from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)
![screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png)


![model](https://user-images.githubusercontent.com/26605804/84588423-ed878900-ae2f-11ea-9994-641ed5323d71.png)

- [ ] Also remember about tests.
## [OPTIONAL] 
* Add screenshot and a live version link of your project to the README.md file.

## [OPTIONAL] 
* What about adding front-end to your project? You don't need to be a designer, but you finished the HTML/CSS part, and already know CSS, SCSS, Bootstrap you can use one!

### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*
*Happy coding * <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png"><img class="emoji" alt="computer" height="25" width="25" src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png"></g-emoji>

RossielCS committed:
Change model user and controllers

- Automatically adds creator id to event

RossielCS committed:
Add bootstrap for views

RossielCS committed:
Add postgres for production

RossielCS committed:
Add rspec and capybara

- New method in events helper
- Add live demo

RossielCS committed:
Add model specs

- add shoulda-matchers gem

RossielCS committed:
Add feature specs

- Update README
- Fix rubocops warnings

RossielCS committed:
Update README

RossielCS committed:
Update events spec features

- Creates an event with a user to show invitations

mosesogwo commented on app/models/user.rb:
```diff
@@ -0,0 +1,29 @@
+  has_many :events, foreign_key: 'creator_id', dependent: :destroy
+  has_many :invitations
+
+  def upcoming_events
+    result = []
+    invitations.each do |e|
+      result << e.event if e.event.date >= DateTime.now.to_date
+    end
+    result
+  end
+
+  def previous_events
+    result = []
+    invitations.each do |e|
+      result << e.event if e.event.date < DateTime.now.to_date
+    end
+    result
```
These should be achieved using scopes
```ruby
scope :upcoming, -> { where('date >= ?', Time.zone.now) }
scope :previous, -> { where('date < ?',  Time.zone.now) }
```

mosesogwo commented on app/models/event.rb:
```diff
@@ -0,0 +1,18 @@
+  scope :upcoming, -> { where('date >= ?', DateTime.now.to_date) }
+  scope :past, -> { where('date < ?', DateTime.now.to_date) }
+
+  def event_attendees
+    users = []
+    invitations.each do |i|
+      users << User.find_by_id(i.user_id)
+    end
+    users
+  end
```
Event attendees should be retrieved by creating  a _has\_many through_ or _has\_and\_belongs\_to\_many_ association with your join model or event model respectively..

![](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/attendee_assoc_requirement.png)

mosesogwo reviewed (changes requested):
**Hi @RossielCS:clap:**
Please see my comments on your code.


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

RossielCS committed:
Change associations

- Remove event_attendees method

RossielCS committed:
Add scopes to user model

RossielCS committed:
Update models' specs

RossielCS committed:
Remove rubocop warnings

wathigo reviewed (approved):
## Approved :heavy_check_mark:  
### Hello @RossielCS :wave: 
Great work implementing the suggested changes.

**You may move on.**

**All the best :four_leaf_clover: 
Simon Wathigo**
